src/core/border_router/routing_manager.cpp

```diff
@@ -730,62 +866,216 @@ void RoutingManager::HandleRouterAdvertisement(const Ip6::Address &aSrcAddress,
+
+    OT_ASSERT(mDiscoveredPrefixNum == 0);
+}
+
+bool RoutingManager::AddDiscoveredPrefix(const Ip6::Prefix &aPrefix,
+                                         bool               aIsOnLinkPrefix,
+                                         uint32_t           aLifetime,
+                                         otRoutePreference  aRoutePreference)
+{
+    OT_ASSERT(IsValidOnLinkPrefix(aPrefix));
+    OT_ASSERT(aLifetime > 0);
+
+    bool added = false;
+
+    for (uint8_t i = 0; i < mDiscoveredPrefixNum; ++i)
+    {
+        ExternalPrefix &prefix = mDiscoveredPrefixes[i];
+        if (aPrefix == prefix.mPrefix && aIsOnLinkPrefix == prefix.mIsOnLinkPrefix)
         {
-            if (mDiscoveredOnLinkPrefix == prefix)
+            prefix.mExpireTime = TimerMilli::GetNow() + GetPrefixExpireDelay(aLifetime);
+            if (prefix.mExpireTime < mDiscoveredPrefixInvalidTimer.GetFireTime())
             {
-                otLogInfoBr("invalidate discovered on-link prefix %s", prefix.ToString().AsCString());
-                mDiscoveredOnLinkPrefixInvalidTimer.Stop();
-                mDiscoveredOnLinkPrefix.Clear();
-                needReevaluate = true;
+                mDiscoveredPrefixInvalidTimer.StartAt(prefix.mExpireTime, 0);
             }
+
+            otLogInfoBr("lifetime of discovered prefix %s is refreshed: %u seconds", aPrefix.ToString().AsCString(),
+                        aLifetime);
+            ExitNow();
         }
-        else
-        {
-            otLogInfoBr("set discovered on-link prefix to %s, valid lifetime: %u seconds",
-                        prefix.ToString().AsCString(), pio->GetValidLifetime());
+    }
 
-            // We keep tracking the latest on-link prefix.
-            mDiscoveredOnLinkPrefixInvalidTimer.Start(GetPrefixExpireDelay(pio->GetValidLifetime()));
-            mDiscoveredOnLinkPrefix = prefix;
+    if (mDiscoveredPrefixNum < kMaxDiscoveredPrefixNum)
+    {
+        ExternalPrefix &newPrefix = mDiscoveredPrefixes[mDiscoveredPrefixNum];
+
+        SuccessOrExit(AddExternalRoute(aPrefix, aRoutePreference));
 
-            // Stop Router Solicitation if we found a valid on-link prefix.
-            // Otherwise, we wait till the Router Solicitation process times out.
+        if (aIsOnLinkPrefix)
+        {
+            // Stop Router Solicitation if we discovered a valid on-link prefix.
+            // Otherwise, we wait till the Router Solicitation process timeouted.
             // So the maximum delay before the Border Router starts advertising
             // its own on-link prefix is 9 (4 + 4 + 1) seconds.
             mRouterSolicitTimer.Stop();
-            needReevaluate = true;
         }
-    }
 
-    if (needReevaluate)
+        newPrefix.mPrefix         = aPrefix;
+        newPrefix.mIsOnLinkPrefix = aIsOnLinkPrefix;
+        newPrefix.mExpireTime     = TimerMilli::GetNow() + GetPrefixExpireDelay(aLifetime);
+        if (newPrefix.mExpireTime < mDiscoveredPrefixInvalidTimer.GetFireTime())
```

Comment:
Can we be sure that `mDiscoveredPrefixInvalidTimer` is started? I think it's meaningless to compare the fire time if `mDiscoveredPrefixInvalidTimer` is not started yet.